### PR TITLE
MatQ Norms

### PR DESCRIPTION
**Pull request by jnsiemer**
This PR implements ..
- [X] norm_sqrd_eucl
- [X] norm_infty
including tests for `MatQ`.

### CODE REVIEW

File: src/rational/mat_q/vector/norm.rs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    ///
+    /// let sqrd_2_norm = vec.norm_sqrd_eucl().unwrap();
+    ///
+    /// assert_eq!(Q::try_from((&14, &1)).unwrap(), sqrd_2_norm);
+    /// ```
+    ///
+    /// Errors and Failures
+    /// - Returns a [`MathError`] of type [`MathError::VectorFunctionCalledOnNonVector`] if
+    /// the given [`MatQ`] instance is not a (row or column) vector.
+    pub fn norm_sqrd_eucl(&self) -> Result<Q, MathError> {
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:42]: I would write "norm_eucl_sqrd", then we have "norm_eucl" and "norm_eucl_sqrd".

File: src/rational/mat_q/vector/norm.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        if !self.is_vector() {
+            return Err(MathError::VectorFunctionCalledOnNonVector(
+                String::from("norm_sqrd_eucl"),
+                self.get_num_rows(),
+                self.get_num_columns(),
+            ));
+        }
+
+        // collect all entries in vector
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:51]: I don't think this comment is needed here

File: src/rational/mat_q/vector/norm.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        let entries = self.collect_entries();
+
+        // sum squared entries in result
+        let mut result = Q::default();
+        for entry in entries {
+            // sets result = result + entry * entry without cloned Q element
+            unsafe { fmpq_addmul(&mut result.value, &entry, &entry) }
+        }
+
+        // TODO: Add sqrt function here
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:61]: Shouldn't this be the squared norm?

File: src/rational/mat_q/vector/norm.rs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+};
+
+impl MatQ {
+    /// Returns the squared Euclidean norm or 2-norm of the given (row or column) vector.
+    ///
+    /// # Example
+    /// ```
+    /// use qfall_math::rational::MatQ;
+    /// use std::str::FromStr;
+    /// # use qfall_math::rational::Q;
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:30]: Why hide this?

File: src/rational/mat_q/vector/norm.rs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    ///
+    /// let vec = MatQ::from_str("[[1/1],[2],[6/2]]").unwrap();
+    ///
+    /// let infty_norm = vec.norm_infty().unwrap();
+    ///
+    /// assert_eq!(Q::try_from((&3, &1)).unwrap(), infty_norm);
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:77]: s.o.

File: src/rational/mat_q/vector/norm.rs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        let mut max = fmpq {
+            num: fmpz(0),
+            den: fmpz(1),
+        };
+        for entry in entries {
+            // compute absolute value of fmpq entry
+            let mut abs_entry = fmpq {
+                num: fmpz(0),
+                den: fmpz(1),
+            };
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:105]: Use defaults now

File: src/rational/mat_q/vector/norm.rs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    pub fn norm_infty(&self) -> Result<Q, MathError> {
+        if !self.is_vector() {
+            return Err(MathError::VectorFunctionCalledOnNonVector(
+                String::from("norm_infty"),
+                self.get_num_rows(),
+                self.get_num_columns(),
+            ));
+        }
+
+        // collect all entries in vector
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:92]: s.o.

File: src/rational/mat_q/vector/norm.rs [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+            if unsafe { fmpq_cmp(&max, &abs_entry) } < 0 {
+                max = abs_entry;
+            }
+        }
+
+        // clone value and ensure that absolute maximum value is absolute
+        let mut result = Q::default();
+        unsafe { fmpq_abs(&mut result.value, &max) }
+
+        Ok(result)
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:117]: Isn't 'max' already absolute?
This seems not necessary.

File: src/rational/mat_q/vector/norm.rs [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+    }
+}
+
+#[cfg(test)]
+mod test_norm_sqrd_eucl {
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:122]: s.o.

File: src/rational/mat_q/vector/norm.rs [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    /// with small entries is calculated correctly
+    #[test]
+    fn row_vector_small_entries() {
+        let vec_1 = MatQ::from_str("[[1]]").unwrap();
+        let vec_2 = MatQ::from_str("[[1,10/1,-1000/10]]").unwrap();
+        let vec_3 = MatQ::from_str("[[1,10,100, 1000]]").unwrap();
+
+        assert_eq!(
+            vec_1.norm_sqrd_eucl().unwrap(),
+            Q::try_from((&1, &1)).unwrap()
```
Comment (Marcel583) [inline src/rational/mat_q/vector/norm.rs:136]: default

**Review by Marcel583 — CHANGES_REQUESTED** [review]

File: src/rational/mat_q/vector/norm.rs (hunk 4 above)
Comment (jnsiemer) [reply src/rational/mat_q/vector/norm.rs]: It's not an import that's important for the function, but I changed it.

File: src/rational/mat_q/vector/norm.rs [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    /// ```
+    /// use qfall_math::rational::MatQ;
+    /// use std::str::FromStr;
+    /// # use qfall_math::rational::Q;
+    ///
+    /// let vec = MatQ::from_str("[[1],[2/1],[6/2]]").unwrap();
+    ///
+    /// let sqrd_2_norm = vec.norm_sqrd_eucl().unwrap();
+    ///
+    /// assert_eq!(Q::try_from((&14, &1)).unwrap(), sqrd_2_norm);
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/rational/mat_q/vector/norm.rs:33]: Because it is not possible to write (14, 1). Otherwise, I would have done that ;)

**Review by Marcel583 — APPROVED** [review]

**Pull request merged**